crates/cli/src/commands/patterns_test.rs

```diff
@@ -276,7 +297,232 @@ pub(crate) async fn run_patterns_test(
+                arg.watch = false; //avoid creating infinite watchers
+                tokio::task::spawn(async move {
+                    let affected_patterns_names =
+                        get_affected_patterns(modified_file_path).await.unwrap();
+
+                    info!(
+                        "[Watch Mode] Patterns changed/affected : {:?}\n[Watch Mode] Re-running test for changed/affected patterns...\n",
+                        affected_patterns_names
+                    );
+
+                    let _ = run_patterns_test(arg, flags, Some(affected_patterns_names)).await;
+                });
+            }
+            Err(error) => {
+                log::error!("Error {error:?}")
+            }
+        }
+    }
+}
+
+async fn get_affected_patterns(modified_file_path: PathBuf) -> Result<Vec<String>> {
+    let modified_file_path_string = modified_file_path.into_os_string().into_string().unwrap();
+
+    let (patterns, _) = resolve_from_cwd(&Source::Local).await?;
+    let affected_patterns = collect_testable_patterns(patterns.clone())
+        .into_iter()
+        .filter(|p| {
+            p.config.path.is_some() && p.config.path.as_ref().unwrap() == &modified_file_path_string
+        })
+        .collect::<Vec<_>>();
+
+    let mut affected_patterns_names = <Vec<String>>::new();
+    for affected_pattern in &affected_patterns {
+        affected_patterns_names.push(affected_pattern.local_name.clone().unwrap());
+    }
+    let mut affected_patterns_dependents_names = <Vec<String>>::new();
+
+    let grit_files = get_grit_files_from_cwd().await?;
+    let current_dir = std::env::current_dir()?;
+    let resolver = GritModuleResolver::new(current_dir.to_str().unwrap());
+
+    for p in patterns {
+        let body = format!("{}()", p.local_name);
+        let lang = PatternLanguage::get_language(&p.body);
+        let grit_files = grit_files.get_language_directory_or_default(lang)?;
+        let rich_pattern = resolver
+            .make_pattern(&body, Some(p.local_name.to_string()))
+            .unwrap();
+        let src = rich_pattern.body;
+        let mut parser = MarzanoGritParser::new()?;
+        let src_tree = parser.parse_file(&src, Some(Path::new(DEFAULT_FILE_NAME)))?;
+        let root = src_tree.root_node();
+        let is_multifile = is_multifile(&root, &grit_files, &mut parser)?;
+
+        let dependents = get_dependents_of_affected_patterns(
+            &grit_files,
+            &src,
+            &mut parser,
+            !is_multifile,
+            &affected_patterns_names,
+        )?;
+
+        for d in dependents {
+            if !affected_patterns_dependents_names.contains(&d) {
+                affected_patterns_dependents_names.push(d);
+            }
+        }
+    }
+
+    for depen in affected_patterns_dependents_names {
+        if !affected_patterns_names.contains(&depen) {
+            affected_patterns_names.push(depen);
+        }
+    }
+    Ok(affected_patterns_names)
+}
+
```

Comment:
Reparsing shouldn't be necessary. If we need additional metadata as the patterns are built in the first place, collect that in the compiler.

Reply:
can you explain this a little more, how to collect the required data in in the compiler? does it happen inside `filter_libs()` in `compiler.rs`?

Reply:
The point is you shouldn't need to reparse any/all patterns to get the dependency tree. Every pattern is already parsed at least one when running the initial test, so you should build the mapping when that happens instead of running again.